dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,221 @@
+    samples_meta = pickle.loads(pickled_meta)
+    return samples_meta
+
+
+def deserialize_batch(buffer: shared_mem.SharedMem, samples):
+    """Deserialize samples from the smem buffer and SampleMeta descriptions.
+
+    Parameters
+    ----------
+    buffer : shared_mem.SharedMem
+        Buffer with serialized sample data
+    samples : List of (idx, SampleMeta) or list of (idx, tuple of SampleMeta).
+        Metadata describing the samples
+
+    Returns
+    -------
+    List of (idx, numpy array) or (idx, tuple of numpy arrays)
+        List of indexed deserialized samples
+    """
+    return [(idx, deserialize_sample(buffer, sample)) for (idx, sample) in samples]
+
+
+class SharedMemChunk:
+    """Simple wrapper around shared memory chunks. Most importantly adds mem_chunk_id used
+    to identify chunks in the communication between parent and worker process
+    (file descriptors cannot serve this purpose easily as the same memory chunk
+    can have different fds in both processes).
+    """
+
+    def __init__(self, mem_chunk_id: str, capacity: int):
+        # mem_chunk_id must be unique among all workers and callbacks in the pool,
+        # used to identify shared memory chunks in the communication between processes
+        self.mem_chunk_id = mem_chunk_id
+        self.shm_chunk = shared_mem.SharedMem.allocate(capacity)
+        self.capacity = capacity
+
+    def resize(self, new_capacity=None):
+        capacity = self.capacity
+        new_capacity = new_capacity or 2 * capacity
+        self.shm_chunk.resize(new_capacity, trunc=True)
+        self.capacity = new_capacity
+
+    def close(self):
+        # Memory fd and mapping will be freed automatically when shm pybind wrapper gets
+        # garbage collected anyway, but you can be so nice to free them as soon as you know
+        # you won't use them
+        self.shm_chunk.close()
+
+
+def _to_numpy(sample):
+    if isinstance(sample, np.ndarray):
+        return sample
+    if types._is_mxnet_array(sample):
+        return sample.asnumpy()
+    if types._is_torch_tensor(sample):
+        if sample.device.type != 'cpu':
+            raise TypeError("GPU tensors are not supported")
+        return sample.numpy()
+    elif isinstance(sample, tensors.TensorCPU):
+        return np.array(sample)
+    raise TypeError(
+        "Unsupported callback return type. Expected numpy array, pytorch or mxnet cpu tensors, "
+        "or list or tuple of them.")
+
+
+def _apply_to_sample(func, sample, *args):
+    """Apply to a sample traversing the nesting of the data (tuple/list).
+
+    Parameters
+    ----------
+    func : callable
+        Function to be applied to every sample data object
+    sample : sample object or any nesting of those in tuple/list
+        Representation of sample
+    """
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(_apply_to_sample(func, part, *args) for part in sample)
```

Comment:
Why not just iterable and return a list.

Reply:
As discussed, we probably need a followup that either allows this or forbids it everywhere.